lightning/src/ln/channelmanager.rs

```diff
@@ -3478,6 +3503,89 @@ impl<ChanSigner: ChannelKeys, M: Deref + Sync + Send, T: Deref + Sync + Send, K:
 	}
 }
 
+/// Used to signal to the ChannelManager persister that the manager needs to be re-persisted to
+/// disk/backups, through `block_until_needs_persist`.
+pub struct PersistenceNotifier {
+	/// Used to signal to the ChannelManager persister that the manager needs to be re-persisted to
+	/// disk/backups. Users won't access the persistence_lock directly, but rather wait on its bool
+	/// using block_until_needs_persist.
+	#[cfg(any(test, feature = "_test_utils"))]
+	pub persistence_lock: Arc<(Mutex<bool>, Condvar)>,
+	#[cfg(not(any(test, feature = "_test_utils")))]
+	persistence_lock: Arc<(Mutex<bool>, Condvar)>,
+}
+
+impl PersistenceNotifier {
+	fn new() -> Self {
+		Self {
+			persistence_lock: Arc::new((Mutex::new(false), Condvar::new())),
+		}
+	}
+
+	/// Used to signal to users when the ChannelManager needs to be persisted to disk. This function
+	/// blocks either for the given Duration or indefinitely until the ChannelManager has updates and
+	/// needs persistence. It returns a bool indicating whether persistence is necessary (which will
+	/// always be true if max_wait is None).
+	pub fn block_until_needs_persist(&self, max_wait: Option<Duration>) -> bool {
+		let current_time = Instant::now();
+		let mut time_to_wait = max_wait;
+		loop {
+			let mutcond = Arc::clone(&self.persistence_lock);
+			let &(ref mtx, ref cvar) = &*mutcond;
+			let mut guard = mtx.lock().unwrap();
+
+			let result = match time_to_wait {
+				Some(wait) => cvar.wait_timeout(guard, wait).unwrap().0,
+				None => cvar.wait(guard).unwrap(),
+			};
+			// Due to spurious wakeups that can happen on `wait_timeout`, here we need to check if the
+			// desired wait time has actually passed, and if not then restart the loop with a reduced
+			// wait time.
+			let elapsed = current_time.elapsed();
+			guard = result;
+			let saved_res = *guard;
+			*guard = false;
+			if saved_res || (max_wait.is_some() && elapsed >= max_wait.unwrap()) {
+				return saved_res;
+			} else if max_wait.is_some() {
+				let remaining_time = max_wait.unwrap().checked_sub(elapsed);
+				if remaining_time.is_some() {
+					time_to_wait = Some(remaining_time.unwrap());
+				} else {
+					// There was actually no time remaining, so just return the result.
+					return saved_res;
+				}
+			}
+		}
+	}
+
+	// Signal to the ChannelManager persister that there are updates necessitating persisting to disk.
+	fn notify(&self) {
```

Comment:
Does a persist-after-every-update make sense ?

* high-throughput, high-value routing node, I guess you will prefer have reliable hardware/system thus preventing state failure and thus passing a longuer `Duration` to avoid latency hit
* low-throughput, high-value, same you will have reliable system and the probability of hitting a failure while having non persisted states are pretty low, thus a reasonable `Duration` should be enough
* low-throughput, low-value, i.e consumer nodes, you will have shitty hardware but optimize for smooth UX thus avoid a disk write after every update ?

Note, we may be glad to provide the option, even if it's hard right now to see if it's going to be heavily used. In anycase, what's your reasoning for `persistent_notifier.notify()` callsites, I guess it's after every newer `MessageSendEvent` but it doesn't sound to properly to bind to it ?

Reply:
> what's your reasoning for persistent_notifier.notify() callsites

Currently it's wherever there's a call to `update_channel` or `add_channel`. But I'm still sort of unclear what the minimal set of callsites is to avoid channel force-closes due to outdated manager.